ICSharpCode.Decompiler/CSharp/Transforms/ExpressionTreeConverter.cs

```diff
@@ -20,51 +20,63 @@
 using System.Collections.Generic;
 using System.Diagnostics;
 using System.Linq;
-using System.Reflection;
-using ICSharpCode.Decompiler.ILAst;
-using ICSharpCode.NRefactory.CSharp;
-using ICSharpCode.NRefactory.PatternMatching;
+using ICSharpCode.Decompiler.CSharp.Resolver;
+using ICSharpCode.Decompiler.CSharp.Syntax;
+using ICSharpCode.Decompiler.CSharp.Syntax.PatternMatching;
+using ICSharpCode.Decompiler.CSharp.TypeSystem;
+using ICSharpCode.Decompiler.IL;
+using ICSharpCode.Decompiler.TypeSystem;
 using Mono.Cecil;
 
 namespace ICSharpCode.Decompiler.CSharp.Transforms
 {
-	public class ExpressionTreeConverter
+	public class ExpressionTreeConverter : DepthFirstAstVisitor, IAstTransform
 	{
+		TransformContext context;
+		TypeSystemAstBuilder astBuilder;
+
+		public ExpressionTreeConverter()
+		{
+		}
+
+		public void Run(AstNode rootNode, TransformContext context)
+		{
+			CSharpResolver resolver = new CSharpResolver(new CSharpTypeResolveContext(context.TypeSystem.Compilation.MainAssembly, null, context.DecompiledTypeDefinition, context.DecompiledMember));
+			astBuilder = new TypeSystemAstBuilder(resolver);
```

Comment:
You are using the same `TypeSystemAstBuilder` in all cases. This is the cause for bugs like this:
```
 379      (-) 		ExpressionTrees.ToCode<bool>(ExpressionTrees.X(), () => new XmlReaderSettings {
      313 (+) 		ExpressionTrees.ToCode<bool>(ExpressionTrees.X(), () => new ExpressionTrees.XmlReaderSettings {
```
1) You are not adding any `UsingScope` to the `CSharpTypeResolveContext`, thus the TSAB will always generate full names.
2) `context.DecompiledTypeDefinition` and `context.DecompiledMember` are not what you think they are. When decompiling a full assembly, they both return null. See https://github.com/icsharpcode/ILSpy/blob/master/ICSharpCode.Decompiler/CSharp/Transforms/TransformContext.cs#L40

In order to properly handle this you will have to do something like this: https://gist.github.com/siegfriedpammer/1d3fe34ec2d7101f63dc1b813d1efafd